rpc_util.go

```diff
@@ -792,15 +797,25 @@ func decompress(compressor encoding.Compressor, d []byte, maxReceiveMessageSize
 // dc takes precedence over compressor.
 // TODO(dfawley): wrap the old compressor/decompressor using the new API?
 func recv(p *parser, c baseCodec, s *transport.Stream, dc Decompressor, m interface{}, maxReceiveMessageSize int, payInfo *payloadInfo, compressor encoding.Compressor) error {
-	d, err := recvAndDecompress(p, s, dc, maxReceiveMessageSize, payInfo, compressor)
+	buf, err := recvAndDecompress(p, s, dc, maxReceiveMessageSize, payInfo, compressor)
 	if err != nil {
 		return err
 	}
-	if err := c.Unmarshal(d, m); err != nil {
+	if err := c.Unmarshal(buf, m); err != nil {
 		return status.Errorf(codes.Internal, "grpc: failed to unmarshal the received message: %v", err)
 	}
 	if payInfo != nil {
-		payInfo.uncompressedBytes = d
+		if p.sharedRecvBufferPool != nil {
+			if len(buf) != 0 {
+				payInfo.uncompressedBytes = make([]byte, len(buf))
+				copy(payInfo.uncompressedBytes, buf)
+			}
+		} else {
+			payInfo.uncompressedBytes = buf
+		}
+	}
+	if p.sharedRecvBufferPool != nil {
+		p.sharedRecvBufferPool.Put(&buf)
```

Comment:
Can this be moved to be inside the above `if p.sharedRecvBufferPool != nil { ... }`?

Reply:
For now, I'll have to settle for that, since I also need to handle the case of `if payInfo != nil { ... }` However, if `sharedRecvBufferPool` is made the default behavior in the future, I believe I can refine the code to be more graceful.